File: string-medium/8.字符串转换整数-atoi.py
#
# @lc app=leetcode.cn id=8 lang=python3
#
# [8] 字符串转换整数 (atoi)
#

import re

# @lc code=start
class Solution:
    # 1082/1082 cases passed (48 ms)
    # Your runtime beats 43.22 % of python3 submissions
    # Your memory usage beats 42.88 % of python3 submissions (14.9 MB)
    def myAtoi(self, s: str) -> int:
        return max(min(int(*re.findall(r"^[\+\-]?\d+", s.lstrip())), 2 ** 31 - 1), -(2 ** 31))


# @lc code=end

class Solution_1:
    def myAtoi(self, str: str) -> int:
        str = str.lstrip()  # 清除左边多余的空格
        num_re = re.compile(r"^[\+\-]?\d+")  # 设置正则规则
        num = num_re.findall(str)  # 查找匹配的内容
        num = int(*num)  # 由于返回的是个列表，解包并且转换成整数
        return max(min(num, 2 ** 31 - 1), -(2 ** 31))  # 返回值

File: string-medium/8.字符串转换整数-atoi.cpp
/*
 * @lc app=leetcode.cn id=8 lang=cpp
 *
 * [8] 字符串转换整数 (atoi)
 */
using namespace std;
#include <string>

// @lc code=start
class Solution {
   public:
    // 1082/1082 cases passed (8 ms)
    // Your runtime beats 34.19 % of cpp submissions
    // Your memory usage beats 75.62 % of cpp submissions (6.9 MB)
    int myAtoi(string str) {
        long res = 0;
        bool instr = false;
        int symbol = 1;
        for (int i = 0; i != str.size(); i++) {
            if (str[i] == ' ' && !instr) {
                continue;
            }
            if (str[i] == '-' && !instr) {
                symbol = -1;
                instr = true;
                continue;
            }
            if (str[i] == '+' && !instr) {
                instr = true;
                continue;
            }
            if (str[i] >= '0' && str[i] <= '9') {
                res = res * 10 + (str[i] - '0');
                instr = true;
                if (symbol * res > INT_MAX)
                    return INT_MAX;
                if (symbol * res < INT_MIN)
                    return INT_MIN;
                continue;
            }
            break;
        }
        return symbol * res;
    }
};
// @lc code=end
